src/extension/extension.ts

```diff
@@ -148,6 +148,7 @@ export function activate(context: vscode.ExtensionContext) {
     Telemetry.sendEvent("compileFileTask");
   });
   let cmdTaskCompileObject = vscode.commands.registerCommand("oradew.compileObjectTask", () => {
+    vscode.commands.executeCommand("workbench.action.tasks.runTask", "oradew: select--sentence");
```

Comment:
What do you think if we still allow the user to manually select a query to execute (sometimes I'd like to execute a query inside some procedure, etc.), but we trigger autoselection only when there is nothing selected (for "Run selected statement" command).

```
createOradewTaskDefinition({
          name: "compile--object",
          params: [
            "compile",
            "--env",
            "${command:oradew.getEnvironment}",
            "--file",
            "${file}",
            "--object",
            "${selectedText}",
            "--line",
            "${lineNumber}",
            "--user",
            "${command:oradew.getUser}",
          ],
        })

``` 
We replace the `"${selectedText}"` parameter in the command with new function, f.e.:
`"${command:oradew.getObjectSelection}",`
that would return your autoselected text if nothing is selected, or current selection in other case.

I think it would be really nice if that could be done...

Reply:
I didn't want to put the changes to this repository yet haha big error. Yes, perfect I will look into that it should be pretty easy to do. I'm about to finnish it. Some questions for you:

I added some code to extension.ts related to code selection, regex and so on, but I don't know if it is the best place to put it, what do you think?

Could you make me a brief summary about how does commands registration and telemetry calls work with an example? I still don't quite understand the flow of command execution in vscode. From top to bottom, something like:

First you register the command here, then you call this, then this get executed etc...

Thanks!

Reply:
Code related to selection functionality could go to `../extension/common/selection.ts` or something like that.

In theory you simply register your command with:
`vscode.commands.registerCommand("oradew.mycommandName", userController.myccommandfunction);`

and add it to `package.json` and it is already available UI within your extension.

If you don't add it to `package.json` it is still available inside tasks but not from UI. (I use `oradew.getEnvironment` for this, f.e.) 

In oradew case is a bit more complicated as `oradew` commands with main functionality is essentially a CLI that can be executed on its own or through extension UI in this case. CLI interface is here [./src/cli/cli.ts](https://github.com/mickeypearce/oradew-vscode/blob/e1d29b939fc79467f7b48861689d2ac00cc2e856/src/cli/cli.ts)

Extension executes this oradew CLI commands through tasks. I don't know if you are familiar, but is a very nice funcionality within vscode to execute extenral commands ([tasks](https://code.visualstudio.com/docs/editor/tasks#_custom-tasks)).

For oradew extension first we register each oradew command as a task with [taskprovider.ts](https://github.com/mickeypearce/oradew-vscode/blob/e1d29b939fc79467f7b48861689d2ac00cc2e856/src/extension/task-provider.ts)
and then assign extension UI command to execute this oradew task with:
```js
let cmdTaskCompileFile = vscode.commands.registerCommand("oradew.compileFileTask", () => {
    vscode.commands.executeCommand("workbench.action.tasks.runTask", "oradew: compile--file");
    Telemetry.sendEvent("compileFileTask");
  });
```

Telemetry is just a simple call to track every command execution for analytics.

As oradew CLI commands are registered as tasks (specific oradew type tasks) you could write your custom tasks with oradew functionality (take a look at [wiki](https://github.com/mickeypearce/oradew-vscode/wiki#3) - 3 - how to customize commands...

If you ask yourself why I complicated my life with tasks etc. it is because it all started with gulp scripts that I then gradually turned into this extension and never had the time nor the vision to refactor into something else. I also quite like the idea that oradew "engine" is build as a CLI and can be used on its own...

Hope I clarified I bit, otherwise just ask....

Reply:
Very helpful, thanks a lot! . Mmm I tried to replace the parameter ${selectedText} of the compile--object command with ${command:oradew.selectCurrentStatement}, but when I do so, the text is properly selected but is not executing anything. Any ideas?